### ISSUE TITLE
Remote mode : Page is set to 1 when I reset a filter

### ISSUE THREAD

**Issue by BenoitAverty** [AllenFang/react-bootstrap-table#1037]
I have a table that I use in remote mode.

If I set the properties like this : 
```
data = <a list of 2 elements>
sizePerPage = 5
options.page = 2
fetchInfo.dataTotalSize=7
```

then it works fine. I see my two elements, and the page number 2 is highlighted.

But when I set a text filter to blank, or a custom filter to blank (I call `filterHandler()` with no parameters inside my custom filter), then the page 1 becomes highlighted in pagination list, and I can't click on it or page 2. The props haven't changed, so I see a mismatch between `props.options.page=2` and `state.currPage=1`.

Do you have a workaround or a fix? I think this is the last bug left in my list and I'm a little bit frustrated ^^

Thanks in advance, and great work with this component, it's very useful.

**Comment by BenoitAverty** [comment]
Also I noticed that it does this when I reset a custom filter, when I remove text from a text filter, but not when I go back to the placeholder in a select filter.

**Comment by AllenFang** [comment]
> Actually I think it's the same behaviour for all kinds of filters

Yes, all the filter or search will go back to page 1, because all the data is different and the data must less than the original data, so it's a default behavior.

> it's just that the problem is hidden for select filters because the data changes and causes new props to be passed to the component.

sorry, I cant get you, any examples or use case? sorry for my pool english...

**Comment by BenoitAverty** [comment]
Sorry, I wasn't very clear because I just wrote all my findings in the same post. I'll try to focus on the essential:

- When a filter changes, the page becomes 1. This makes sense for the table in normal mode because the data changes.
- In remote mode, this is problematic because it causes a mismatch between the page that's in the table state, and the page that's in the props.options object. The result is that clicking on the links doesn't work anymore. I think the reset should be made to options.page instead of the first page index (if it's specified).
- Another problem if that the page is reset to one even if the data doesn't change. For example, if the filter was already empty and it calls the handler again, then the page is reset to one even though nothing has changed. 

I hope it's more clear, don't hesitate to ask if something isn't clear

**Comment by AllenFang** [comment]
> In remote mode, this is problematic because it causes a mismatch between the page that's in the table state, and the page that's in the props.options object. The result is that clicking on the page links doesn't work anymore. I think the reset should be made to options.page instead of the first page index (if it's specified).

I'll check this out, some questions is your `remote` mode is [alternative](https://github.com/AllenFang/react-bootstrap-table/blob/master/examples/js/remote/remote-alternative.js#L9)?, and did you change the `options.page` in the `onFilterChange`? BTW,** I think you may be right**, but I can't reproduce this issue currently so not sure very well


So I wanna check your code, could you help me to create a simple and minimal example which can describe your second case? sorry for I'm busy on my job so it's really helpful for me to reproduce issue easily, take your time, you can follow [these](https://github.com/AllenFang/react-bootstrap-table/tree/master/examples/js/remote) example to create a your case, Thanks 👍

**Comment by BenoitAverty** [comment]
I'm not using the alternative remote. My table is not editable, I only display data and sort/filter/paginate it.

I have made a minimal repo to reproduce : https://github.com/BenoitAverty/example-bootstraptable

Just checkout / npm install / open in chrome (or other modern browser).

Steps to reproduce : 
 - Open the page
 - Click on page 2
 - Change the filter (value doesn't matter because the handler does nothing)

What will happen is : 
 - The table component will have its state.currPage equals to 1 (this is easy to see with something like react-dev-tools chrome extension)
 - The table component will still have its props.options.page equals to 2 (because the parent compoenent didn't re-render)
 - You will not be able to change pages

What should happen in my opinion:
 - The state.currPage should stay equals to props.options.page. It's the responsibility of the user to make sure options.page is consistent with its data

This of course is only true in remote mode. In normal mode, changing the page to 1 would behave as normal because the data would change and options.page is only the default page.

Thanks in advance.

**Comment by AllenFang** [comment]
@BenoitAverty, it's awesome, thanks for your repo!!

**Comment by AllenFang** [comment]
@BenoitAverty, after my investigation for your code, I think I need to explain `remote` firstly,
Actually, **`remote` means you are supposed to handle all the data change** and then render your table, so this issue is because there's no any change when filtering but table itself have change some state so that cause a different situation between your state and table state.

Actually, I just write a example which have a remote search and pagination and it goes work well, so I just wanna to ask why you did nothing when filtering? did you have any concern, I just want to know, maybe there have some feature that I can support,

Thanks

**Comment by BenoitAverty** [comment]
I do nothing in this example, but this is just because making a real example would be too long, and unfortunately the project I'm working on isn't open-source.

However, there are several case where changing a filter wouldn't change the data : Imagine, for example the folllwing data : 
```
[
 {key: 1, date: 2017-02-19},
 {key: 2, date: 2017-02-20},
 {key: 3, date: 2017-02-21},
 {key: 4, date: 2017-02-22},
 {key: 5, date: 2017-02-23},
 {key: 6, date: 2017-02-24},
 {key: 7, date: 2017-02-25},
 {key: 8, date: 2017-02-26},
]
```

and a date filter that is set to " after 2017-01-01". In that case the filter doesn't actually exclude any data.
If you go to page 2 then clear this date filter, then the data won't change : the 3 rows of page 2 are still displayed. But the page will still be reset to 1 and cause the bug, even though nothing should have changed because the data is exactly the same.

**Comment by AllenFang** [comment]
ya, I understand it, but if you do nothing in the filtering, it's not the expectation for `remote mode`, so I suggest you just to set the complete or the origin data again(Don't filtering, just change state with same data).

Actually, if your filter dont change data always, you can use [remote](https://github.com/AllenFang/react-bootstrap-table/blob/master/examples/js/remote/remote-alternative.js#L9) to disable the column filtering with remote.

Please let me know any concern here.

**Comment by BenoitAverty** [comment]
I don't understand why not changing the data is unexpected. As I said, there are plenty of cases where changing a filter would not change the data.
Another axample would be to type a text in a text filter, but all the data contains that test. It's very counter-intuitive to go back to page 1 in that case.
Forcing a re-render of the component would maybe work (not sure), but it would be highly inefficient and fixing this in the library makes more sense.

I updated the repo to make the text filter functional. You can see that the filter works : for example, typing "3" will only show the "value3" line.

However, you can see the bug I'm talking about by following these steps : 
 - Clear the filter : all lines should show
 - Go to page 2
 - type "val" in the filter.

The page will go back to 1 and be stuck. I don't think this is the correct behavior. I was able to find a workaround by setting the page to "1" in the "handleFilter" function, but this is not the behavior I want.

Do you want me to open a PR for this bug ?

**Comment by AllenFang** [comment]
Thanks for your investigation further,  
> Another example would be to type a text in a text filter, but all the data contains that test. It's very counter-intuitive to go back to page 1 in that case.

For me, no matter remote or not, jump to page 1 is truth case, no one argu this before, and why I do this, you can check jqGrid and DataTable, they are all familiar table component and they have same behavior with `react-bootstrap-table`, even if your data is fully match search pattern. So I just follow most case to design this functionality and I think I wont to change this behavior. But if there're too many people think this is a strange behavior, I'll do it immediately.

After I run your code and I figure out the root cause, The problem is **I should let you know the page change to 1, but I didn't, I just change it [internal](https://github.com/AllenFang/react-bootstrap-table/blob/master/src/BootstrapTable.js#L763) and cause you slice the wrong data**, 

I'll find a solution to fix it, but there're couple thing I need to concern

1. I wont change the functionality for changing to page1
2. I'll change to page by calling internal `handlePagination` and this function will control the remote mode to let you receive the `onPageChange` event so that you must have a correct page number.... so you will not slice wrong data anymore.
3. Maybe I'll provide a props to configure to not change to page 1 when enable remote with search/filtering

Thanks :)

**Comment by BenoitAverty** [comment]
Ok, calling the page change callback will solve the problem at least, there won't be a bug anymore.

I still think the table shouldn't change the page by itself when in remote mode, but it's your choice and I'll respect it.

Thanks for your time and your work on this component.

**Comment by AllenFang** [comment]
This enhancement will be done after `v3.0.0` release, so need some time, sorry, I'm busy.

**Comment by BenoitAverty** [comment]
No problem, Do it when it's convenient :)

Benoit AVERTY
Consultant

Zenika Rennes
2 bis Place St-Mélaine - 35000 RENNES

Standard : +33(0)2 22 66 61 50
Fax : +33(0)2 22 66 61 51


2017-02-21 17:02 GMT+01:00 Allen <notifications@github.com>:

**Comment by dekelb** [comment]
@BenoitAverty I must say I'm with @AllenFang here.

When a filter is applied, you don't know how exactly the new data looks like and if it's the same data you already have in you current page or not, so once a filter is applied - you should go back to page 1.

Even if you have rows X,Y in your current page, and you applied some filter, and after the filter you still get from your server rows X, Y - no one guarantee that it's exactly the same page.

**Comment by BenoitAverty** [comment]
I agree with that, except when you clear an already empty filter. Either
way, when the table is in remote mode my opinion is that it shouldn't
change anything on its own. But as I said, I have a workaround for the bug
and I respect the other opinion so there's no problem :)

**Comment by dekelb** [comment]
Can you explain a bit more about "already empty filter"? If the filter is empty (nothing is filtered) - how exactly do you clear it?
Or you mean that the filter didn't filter anything? (didn't affect the results).

**Comment by BenoitAverty** [comment]
For the "clearing" use case, it's a custom filter with a date field. I have
a button which clears the filter. It can also happen if you quickly type
then remove a character in a text filter.

**Comment by 9005** [comment]
@BenoitAverty : I am facing similar issue as you said. What;s fix. I do text filter and use onFilterChange method and page Number reset to 1. Need help

**Comment by 9005** [comment]
@BenoitAverty  I am doing API call on page change, So if i do call back to onPagechanage method then i have to do API call again.
